### DISCUSSION TITLE
Welcome to PiTender Discussions!

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiTender#4] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ardycloud** [comment]
Hi have installed PiTender on my Pi Zero W.
The auto installer worked without problems.
But unfortunately every time I click on something that activates a pump I get this error message: https://i.imgur.com/MKDWGkc.png
But the pumps are activated.
But I find this GitHub project really good especially because of the web GUI and the autoinstaller.

Greetings from Austria

**Comment by nebhead** [comment]
Hey @ardycloud - Glad you found your way to the project!  Can you log into your Pi and check the `/home/pi/pitender/logs` folder.  I'm specifically interesting in seeing the webapp logs.  I suspect there is a python error somewhere, perhaps something isn't supported on your Pi OS for some reason.  Speaking of, what version of Raspberry Pi OS are you running on the Pi Zero?  

Also, I have been to Austria a couple of times in my life and it has to be one of the most beautiful places I have ever seen.

**Reply by ardycloud** [reply]
Hi nebhead,

thank you for your quick reply.
In the attachment you will find the two log files that are not empty.
As image I used the latest from the Raspian website.
It is a Light without Desktop IMG.

Another quick question: a small 0.96 inch with rotary endcoder is not possible to use as well or? would be cool to use standalone without mobile / tablet.

[events.log](https://github.com/nebhead/PiTender/files/6481533/events.log)
[webapp.err.log](https://github.com/nebhead/PiTender/files/6481534/webapp.err.log)

**Reply by nebhead** [reply]
Thanks!  I'll take a look and see if I can reproduce the error on my end. 

In the webapp.err.log I see the following error:

```
  File "/home/pi/pitender/templates/work.html", line 5, in block 'head'
    {% if "cancel" in action %}
jinja2.exceptions.UndefinedError: 'action' is undefined
```
If I don't get to this right away, maybe try to add the action variable on the following lines 124, 372, 392 in app.py:

```
			return render_template('work.html', drink_name=drink_name, action="default")
```

Then reboot and try again.

**Reply by ardycloud** [reply]
Hi,

jetzt klappt es :) 
Gibt es in Zukunft auch noch die möglichkeit selber eigene Zutaten hinzuzufügen?

**Reply by nebhead** [reply]
Wonderful!  Glad to hear it's working.  I'll make those changes and push an update.  

To add a new ingredient, go to the recipe builder page and in the ingredients section, you can add a new ingredient there.  

Take care and enjoy!!

**Reply by ardycloud** [reply]
Hi,

have now tried to create a new drink.
When I create this drink, it is not displayed in the drinks list.
I also wanted to create my own image for it.
I uploaded the image to the folder.
Image upload is unfortunately not possible.
Enclosed again the logfile.
Please let me know what formats the images have to be displayed well.

Kind regards
[webapp.err.log](https://github.com/nebhead/PiTender/files/6490094/webapp.err.log)

**Reply by nebhead** [reply]
Yep, I haven't implemented image upload yet - but will look into it for future versions.  Not sure what the issue is with loading the drink images that is showing in the log file.  I might try to duplicate this myself.  Images are stored in `pitender/static/img/drinks/` and are either jpg or png (basically anything your browser can display).  If you want, you can start with the drink images that exist and add a new image of the same size.

**Reply by ardycloud** [reply]
Hi, 

I found the problem.
The problem was, that the ingredients weren't defined. The use of custom images for the drinks is working too.
Also I wanted to ask what has to be changed in the code in order to use more pumps.

Kind regards
Rene

**Reply by nebhead** [reply]
Hey @ardycloud - Glad to hear you found the issue.  I really need to implement some better feedback in the tool so it's more user friendly. :)  

As for adding pumps, it's probably possible, but would need to modify control.py and the settings.json to add the pump numbers to the available pumps.  How many pumps did you need?

**Reply by ardycloud** [reply]
Hi,
currently I would like to control 10 or if possible 12 pumps with it.
So if then only add the pumps in the control.py and the settings.json ?
Can I use all GPIO's that are shown in the admin page?

**Reply by nebhead** [reply]
Yeah, technically speaking, it should be able to handle as many pumps as there are available GPIOs (of course, you'll have to consider your power-load for the pumps and power-supply, but that's maybe out of scope).   I'll try to give this a test in a simulated environment a little later today and see if I can get it to work.

**Reply by nebhead** [reply]
OK, I just did a quick simulated test and it appears that no code really needs to be altered to add more pumps.  Instead, you can edit the `settings.json` to add additional pumps, then restart.  

Below is an example, adding pump_09 to the list.  Just take care to not duplicate GPIO pin numbers. 

```
{
    "assignments": {
        "pump_02": 27,
        "pump_03": 22,
        "pump_01": 17,
        "pump_06": 25,
        "pump_07": 2,
        "pump_04": 23,
        "pump_05": 24,
        "pump_08": 3,
        "pump_09": 26
    },
    "inventory": {
        "pump_02": "vodka",
        "pump_03": "whiskey",
        "pump_01": "rum",
        "pump_06": "tequila",
        "pump_07": "gin",
        "pump_04": "coke",
        "pump_05": "oj",
        "pump_08": "cranberry",
        "pump_09": "kahlua"
    },
    "flowrate": 85
}
```

**Reply by ardycloud** [reply]
Hi,

I have tested during cleaning comes at pump 9 and 10. the message that the drink is prepared but no relay is controlled.
At pump 11. and 12. the relay or GPIO is used by the 1. and 2 relay.

**Reply by nebhead** [reply]
Hey, I'm not sure if I fully understand.  Can you post your `settings.json` file here?  Can you send me the steps to reproduce and which relays you would expect to be turned on?  

How many relays do you have connected?  Do you have power/gnd connected for all relays/modules?

Also, did you reboot after modifying the `settings.json` file?

**Reply by ardycloud** [reply]
Hi,

I use two relay boards one with 8 channel and one with 4 channel.
Both have their own power supply.
The board with 8 channel is from 1-8 and the one with 4 channel is 9-12.
After changing the .json file I restarted the Pi.
e.g. Pump 9 is to control the 4-channel board Relay 1.
But it controls the 8 channel board relay 1.
Sorry, I don't know much about adapting code.
With the project I try to deal with it a little more

`{
    "assignments": {
        "pump_02": 27,
        "pump_03": 22,
        "pump_01": 17,
        "pump_06": 25,
        "pump_07": 2,
        "pump_04": 23,
        "pump_05": 24,
        "pump_08": 3,
        "pump_09": 18,
        "pump_10": 19,
        "pump_11": 20,
        "pump_12": 21
    },
    "inventory": {
        "pump_02": "vodka",
        "pump_03": "whiskey",
        "pump_01": "rum",
        "pump_06": "tequila",
        "pump_07": "gin",
        "pump_04": "coke",
        "pump_05": "oj",
        "pump_08": "cranberry",
        "pump_09": "kahlua",
        "pump_10": "",
        "pump_11": "",
        "pump_12": ""
    },
    "flowrate": 40
}`

**Reply by nebhead** [reply]
OK, I see what's happening.  Thanks for sharing all of your details!  I copied your json data and pasted into my simulation.  

So, it seems like the drink dispensing code seems to work OK with the extra pumps in the configuration.  However, the _cleaning_ function has a different method of running pumps and probably doesn't recognize pump numbers higher than 8.  Thus I will likely need to edit the code to make this work.  However, for the time being, you could still use this to mix drinks - just won't be able to clean it properly using this function. :)

**Reply by nebhead** [reply]
Alright, I decided to go ahead an update the source with these changes to allow more pumps with the cleaning function.  Latest should have the changes from this commit (https://github.com/nebhead/PiTender/commit/16f2a0bb0b2634985d679f0eb555dece4cbb9448)

**Reply by ardycloud** [reply]
Hi,

now pump 9,10,11,12 works.
Thanks for the quick fix :)

**Reply by nebhead** [reply]
Outstanding!  You're welcome. :)

**Comment by davrad3** [comment]
Hello,
I'll try to build the Smart Bartender, but I can't find the 3D Files.
Can you guys help me?
Thanks in advance!
Greetings
